# Bag 1: Import Library

In [21]:
import pandas as pd               # Baca data, olah data, grouping, merging
import numpy as np                # Kalkulasi, array, matriks
import matplotlib.pyplot as plt   # Grafik sederhana
import seaborn as sns             # Grafik tinggi = peta, heatmap

# Bag 2: Baca Data

In [22]:
df = pd.read_csv('ds_salaries.csv')
df.head(10) # tampilkan 10 data teratas

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
0,2023,SE,FT,Principal Data Scientist,80000,EUR,85847,ES,100,ES,L
1,2023,MI,CT,ML Engineer,30000,USD,30000,US,100,US,S
2,2023,MI,CT,ML Engineer,25500,USD,25500,US,100,US,S
3,2023,SE,FT,Data Scientist,175000,USD,175000,CA,100,CA,M
4,2023,SE,FT,Data Scientist,120000,USD,120000,CA,100,CA,M
5,2023,SE,FT,Applied Scientist,222200,USD,222200,US,0,US,L
6,2023,SE,FT,Applied Scientist,136000,USD,136000,US,0,US,L
7,2023,SE,FT,Data Scientist,219000,USD,219000,CA,0,CA,M
8,2023,SE,FT,Data Scientist,141000,USD,141000,CA,0,CA,M
9,2023,SE,FT,Data Scientist,147100,USD,147100,US,0,US,M


In [23]:
df.tail(10) # tampilkan 10 data terakhir

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
3745,2021,SE,FT,Director of Data Science,168000,USD,168000,JP,0,JP,S
3746,2021,MI,FT,Data Scientist,160000,SGD,119059,SG,100,IL,M
3747,2021,MI,FT,Applied Machine Learning Scientist,423000,USD,423000,US,50,US,L
3748,2021,MI,FT,Data Engineer,24000,EUR,28369,MT,50,MT,L
3749,2021,SE,FT,Data Specialist,165000,USD,165000,US,100,US,L
3750,2020,SE,FT,Data Scientist,412000,USD,412000,US,100,US,L
3751,2021,MI,FT,Principal Data Scientist,151000,USD,151000,US,100,US,L
3752,2020,EN,FT,Data Scientist,105000,USD,105000,US,100,US,S
3753,2020,EN,CT,Business Data Analyst,100000,USD,100000,US,100,US,L
3754,2021,SE,FT,Data Science Manager,7000000,INR,94665,IN,50,IN,L


In [24]:
df.sample() # mengambil secara acak 1 data

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
2005,2022,SE,FT,Machine Learning Engineer,246000,USD,246000,US,100,US,M


# Bag 3: Eksplorasi Data

In [25]:
df.shape

(3755, 11)

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


Tidak ada missing value

In [27]:
# Kategori unik experience_level
df['experience_level'].unique()

array(['SE', 'MI', 'EN', 'EX'], dtype=object)

In [28]:
# Kategori unik employment_type
df['employment_type'].unique()

array(['FT', 'CT', 'FL', 'PT'], dtype=object)

# Bag 4: Cleaning Data

## Missing Value

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [30]:
df.isnull().sum() # menghitung jumlah data yang kosong per kolomnya

,0
work_year,0
experience_level,0
employment_type,0
job_title,0
salary,0
salary_currency,0
salary_in_usd,0
employee_residence,0
remote_ratio,0
company_location,0


Tidak ada missing value

## Data Duplikat

In [31]:
df[df.duplicated()]

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
115,2023,SE,FT,Data Scientist,150000,USD,150000,US,0,US,M
123,2023,SE,FT,Analytics Engineer,289800,USD,289800,US,0,US,M
153,2023,MI,FT,Data Engineer,100000,USD,100000,US,100,US,M
154,2023,MI,FT,Data Engineer,70000,USD,70000,US,100,US,M
160,2023,SE,FT,Data Engineer,115000,USD,115000,US,0,US,M
...,...,...,...,...,...,...,...,...,...,...,...
3439,2022,MI,FT,Data Scientist,78000,USD,78000,US,100,US,M
3440,2022,SE,FT,Data Engineer,135000,USD,135000,US,100,US,M
3441,2022,SE,FT,Data Engineer,115000,USD,115000,US,100,US,M
3586,2021,MI,FT,Data Engineer,200000,USD,200000,US,100,US,L


In [32]:
df.duplicated().sum()

np.int64(1171)

In [33]:
df[df.duplicated()].sort_values(by=['work_year','experience_level','employment_type','job_title','salary'])

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
3120,2020,EN,FT,Data Engineer,1000000,INR,13493,IN,100,IN,L
3586,2021,MI,FT,Data Engineer,200000,USD,200000,US,100,US,L
3709,2021,MI,FT,Data Scientist,76760,EUR,90734,DE,50,DE,L
2041,2022,EN,FT,Data Analyst,50000,USD,50000,US,50,US,L
3272,2022,EN,FT,Data Engineer,40000,GBP,49253,GB,100,GB,M
...,...,...,...,...,...,...,...,...,...,...,...
1560,2023,SE,FT,Research Scientist,210000,USD,210000,US,0,US,M
1706,2023,SE,FT,Research Scientist,210000,USD,210000,US,0,US,M
956,2023,SE,FT,Research Scientist,250000,USD,250000,US,0,US,M
1751,2023,SE,FT,Research Scientist,250000,USD,250000,US,0,US,M


Ada 1171 baris yang terduplikat satu sama lain.

Solusi: Dihapus

In [34]:
df.drop_duplicates(inplace=True) # hapus di dataset itu

In [35]:
df2 = df.drop_duplicates() # simpan dataset yang sudah bersih dari duplikat df2

# Bag 5: Stat Des

In [37]:
df2.describe() # stat des kolom numerik

,work_year,salary,salary_in_usd,remote_ratio
count,2584.000000,2.584000e+03,2584.000000,2584.000000
mean,2022.301084,2.103653e+05,133409.280186,50.483746
std,0.749179,8.080375e+05,67136.837329,48.163707
min,2020.000000,6.000000e+03,5132.000000,0.000000
25%,2022.000000,9.000000e+04,84975.000000,0.000000
50%,2022.000000,1.346300e+05,130000.000000,50.000000
75%,2023.000000,1.825625e+05,175000.000000,100.000000
max,2023.000000,3.040000e+07,450000.000000,100.000000


In [38]:
df2.describe(include = 'object') # kolom yang bertipe string/kategorik

,experience_level,employment_type,job_title,salary_currency,employee_residence,company_location,company_size
count,2584,2584,2584,2584,2584,2584,2584
unique,4,4,93,20,78,72,3
top,SE,FT,Data Engineer,USD,US,US,M
freq,1554,2547,598,2107,1893,1929,2028


# Bag 6: Mapping Kategori

## Mapping kategori experience_level

In [39]:
df2['experience_level'].unique()

array(['SE', 'MI', 'EN', 'EX'], dtype=object)

* SE = SENIOR
* MI = MIDDLE
* EN = ENTRY
* EX = EXECUTIVE

In [40]:
# Buat dictionary
dict_exp_level = {
    'SE': 'SENIOR',
    'MI': 'MIDDLE',
    'EN': 'ENTRY',
    'EX': 'EXECUTIVE'
}

In [41]:
df2['experience_level'] = (df2['experience_level'].map(dict_exp_level))

In [42]:
df2['experience_level'].unique()

array(['SENIOR', 'MIDDLE', 'ENTRY', 'EXECUTIVE'], dtype=object)

## Mapping kategori employment_type

In [44]:
df2['employment_type'].unique()

array(['FT', 'CT', 'FL', 'PT'], dtype=object)

In [45]:
dict_emp_type = {
    'FT': 'FULL TIME',
    'CT': 'CONTRACT',
    'FL': 'FREELANCE',
    'PT': 'PART TIME'
}

In [46]:
df2['employment_type'] = (df2['employment_type'].map(dict_emp_type))

In [47]:
df2['employment_type'].unique()

array(['FULL TIME', 'CONTRACT', 'FREELANCE', 'PART TIME'], dtype=object)

## Mapping kategori remote_ratio

In [48]:
df2['remote_ratio'].unique()

array([100,   0,  50])

In [49]:
dict_remote_ratio = {
    0 : 'ONSITE',
    50 : 'HYBRID',
    100 : 'REMOTE'
}

In [50]:
# Membuat kolom baru kategori_remote
df2['remote_category'] = (df['remote_ratio'].map(dict_remote_ratio))
df2

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size,remote_category
0,2023,SENIOR,FULL TIME,Principal Data Scientist,80000,EUR,85847,ES,100,ES,L,REMOTE
1,2023,MIDDLE,CONTRACT,ML Engineer,30000,USD,30000,US,100,US,S,REMOTE
2,2023,MIDDLE,CONTRACT,ML Engineer,25500,USD,25500,US,100,US,S,REMOTE
3,2023,SENIOR,FULL TIME,Data Scientist,175000,USD,175000,CA,100,CA,M,REMOTE
4,2023,SENIOR,FULL TIME,Data Scientist,120000,USD,120000,CA,100,CA,M,REMOTE
...,...,...,...,...,...,...,...,...,...,...,...,...
3750,2020,SENIOR,FULL TIME,Data Scientist,412000,USD,412000,US,100,US,L,REMOTE
3751,2021,MIDDLE,FULL TIME,Principal Data Scientist,151000,USD,151000,US,100,US,L,REMOTE
3752,2020,ENTRY,FULL TIME,Data Scientist,105000,USD,105000,US,100,US,S,REMOTE
3753,2020,ENTRY,CONTRACT,Business Data Analyst,100000,USD,100000,US,100,US,L,REMOTE


# Bag 7: EDA

In [51]:
df2['experience_level'].value_counts()

,count
experience_level,
SENIOR,1554
MIDDLE,664
ENTRY,270
EXECUTIVE,96


Ada berapa employment fulltime?

In [53]:
df2['employment_type'].value_counts()

,count
employment_type,
FULL TIME,2547
PART TIME,17
CONTRACT,10
FREELANCE,10


In [54]:
df['employment_type'].value_counts(normalize=True) # cari persentasenya

,proportion
employment_type,
FT,0.985681
PT,0.006579
CT,0.003870
FL,0.003870


In [52]:
df2

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size,remote_category
0,2023,SENIOR,FULL TIME,Principal Data Scientist,80000,EUR,85847,ES,100,ES,L,REMOTE
1,2023,MIDDLE,CONTRACT,ML Engineer,30000,USD,30000,US,100,US,S,REMOTE
2,2023,MIDDLE,CONTRACT,ML Engineer,25500,USD,25500,US,100,US,S,REMOTE
3,2023,SENIOR,FULL TIME,Data Scientist,175000,USD,175000,CA,100,CA,M,REMOTE
4,2023,SENIOR,FULL TIME,Data Scientist,120000,USD,120000,CA,100,CA,M,REMOTE
...,...,...,...,...,...,...,...,...,...,...,...,...
3750,2020,SENIOR,FULL TIME,Data Scientist,412000,USD,412000,US,100,US,L,REMOTE
3751,2021,MIDDLE,FULL TIME,Principal Data Scientist,151000,USD,151000,US,100,US,L,REMOTE
3752,2020,ENTRY,FULL TIME,Data Scientist,105000,USD,105000,US,100,US,S,REMOTE
3753,2020,ENTRY,CONTRACT,Business Data Analyst,100000,USD,100000,US,100,US,L,REMOTE


# Bag 8: Grouping Data

Pivot table (ga ada jadi kolom)

Berapakah rata-rata gaji berdasarkan experience_level?

In [55]:
df2.groupby('experience_level')['salary'].mean()

,salary
experience_level,
ENTRY,199203.914815
EXECUTIVE,252675.343750
MIDDLE,276429.332831
SENIOR,181462.650579


Berapa gaji maksimum berdasarkan employment_type?

In [56]:
df2.groupby('employment_type')['salary'].max()

,salary
employment_type,
CONTRACT,416000
FREELANCE,2400000
FULL TIME,30400000
PART TIME,400000


Contoh groupby gabungan

In [58]:
df2.groupby(['experience_level','employment_type'])[['salary','salary_in_usd']].mean()

salary  salary_in_usd
experience_level employment_type                              
ENTRY            CONTRACT          64500.000000   65234.500000
                 FREELANCE         75000.000000   75000.000000
                 FULL TIME        208954.654762   74564.626984
                 PART TIME         60677.428571   38885.000000
EXECUTIVE        CONTRACT         416000.000000  416000.000000
                 FULL TIME        250956.136842  188710.610526
MIDDLE           CONTRACT          78600.000000   78600.000000
                 FREELANCE        515511.000000   41615.600000
                 FULL TIME        276604.488479  102742.781874
                 PART TIME        169666.666667   42561.000000
SENIOR           CONTRACT          97500.000000   97500.000000
                 FREELANCE         53333.333333   53333.333333
                 FULL TIME        181819.211750  154165.019367

Berapa maximum salary dan salary_in_usd per company_size dan remote_ratio

In [61]:
df2.groupby(['company_size','remote_ratio'])[['salary','salary_in_usd']].max()

salary  salary_in_usd
company_size remote_ratio                         
L            0              4900000         423834
             50            11000000         423000
             100           30400000         412000
M            0              2500000         450000
             50             1800000         173000
             100            6600000         376080
S            0               720000         275000
             50             8500000         165000
             100            4450000         416000

In [62]:
# Mau format tabel (tidak di merge)
df2.groupby(['company_size','remote_ratio'])[['salary','salary_in_usd']].max().reset_index()

,company_size,remote_ratio,salary,salary_in_usd
0,L,0,4900000,423834
1,L,50,11000000,423000
2,L,100,30400000,412000
3,M,0,2500000,450000
4,M,50,1800000,173000
5,M,100,6600000,376080
6,S,0,720000,275000
7,S,50,8500000,165000
8,S,100,4450000,416000


In [63]:
df.groupby('experience_level').agg({
    'salary': 'max',
    'work_year': 'mean'
})

,salary,work_year
experience_level,,
EN,6600000,2021.940741
EX,6000000,2022.291667
MI,30400000,2022.150602
SE,7500000,2022.428571


Tampilkan data per job_title, tampilkan rata-rata, total, min dari salary_in_usd, dan jumlah (count/size) dari experience_level

In [64]:
df2.groupby('job_title').agg({
    'salary_in_usd': ['mean', 'count', 'min'],
    'experience_level': 'count'
})

salary_in_usd               experience_level
                                        mean count     min            count
job_title                                                                  
3D Computer Vision Researcher   21352.250000     4    5409                4
AI Developer                   136666.090909    11    6304               11
AI Programmer                   55000.000000     2   40000                2
AI Scientist                   110120.875000    16   12000               16
Analytics Engineer             150151.747253    91    7500               91
...                                      ...   ...     ...              ...
Research Engineer              165909.393939    33   60000               33
Research Scientist             153486.369231    65   23000               65
Software Data Engineer          62510.000000     2   50000                2
Staff Data Analyst              15000.000000     1   15000                1
Staff Data Scientist           105000.000000     1  105000                1

[93 rows x 4 columns]

# Bag 9: Sorting

In [65]:
df2.sort_values(by='salary', ascending=False) # urutan data berdasarkan gaji terendah

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size,remote_category
3669,2021,MIDDLE,FULL TIME,Data Scientist,30400000,CLP,40038,CL,100,CL,L,REMOTE
3574,2021,MIDDLE,FULL TIME,BI Data Analyst,11000000,HUF,36259,HU,50,US,L,HYBRID
3646,2020,MIDDLE,FULL TIME,Data Scientist,11000000,HUF,35735,HU,50,HU,L,HYBRID
3475,2021,MIDDLE,FULL TIME,ML Engineer,8500000,JPY,77364,JP,50,JP,S,HYBRID
2966,2022,SENIOR,FULL TIME,Lead Machine Learning Engineer,7500000,INR,95386,IN,50,IN,L,HYBRID
...,...,...,...,...,...,...,...,...,...,...,...,...
2589,2022,SENIOR,FULL TIME,Product Data Scientist,8000,USD,8000,IN,100,SG,L,REMOTE
3681,2020,MIDDLE,FULL TIME,Data Analyst,8000,USD,8000,PK,50,PK,L,HYBRID
2933,2022,MIDDLE,CONTRACT,Analytics Engineer,7500,USD,7500,BO,50,BO,M,HYBRID
573,2023,ENTRY,FULL TIME,Autonomous Vehicle Technician,7000,USD,7000,GH,0,GH,S,ONSITE


In [68]:
df2.sort_values(by=['work_year','experience_level'], ascending=True)

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size,remote_category
3119,2020,ENTRY,FULL TIME,Data Engineer,1000000,INR,13493,IN,100,IN,L,REMOTE
3138,2020,ENTRY,FULL TIME,Data Analyst,20000,EUR,22809,PT,100,PT,M,REMOTE
3163,2020,ENTRY,FULL TIME,Data Scientist,43200,EUR,49268,DE,0,DE,S,ONSITE
3446,2020,ENTRY,FULL TIME,Data Engineer,48000,EUR,54742,PK,100,DE,L,REMOTE
3514,2020,ENTRY,FULL TIME,Big Data Engineer,70000,USD,70000,US,100,US,L,REMOTE
...,...,...,...,...,...,...,...,...,...,...,...,...
1805,2023,SENIOR,FULL TIME,Data Architect,110000,USD,110000,US,100,US,M,REMOTE
1808,2023,SENIOR,FULL TIME,Data Engineer,236000,USD,236000,US,100,US,M,REMOTE
1809,2023,SENIOR,FULL TIME,Data Engineer,182000,USD,182000,US,100,US,M,REMOTE
1814,2023,SENIOR,FULL TIME,Machine Learning Engineer,261500,USD,261500,US,0,US,L,ONSITE


Urutan berdasarkan company_location, company_size, (a-z)

In [70]:
df.sort_values(by=['company_location','company_size'], ascending=True)

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
3607,2020,MI,FT,Lead Data Scientist,115000,USD,115000,AE,0,AE,L
3408,2022,SE,FT,Machine Learning Engineer,120000,USD,120000,AE,100,AE,S
3409,2022,SE,FT,Machine Learning Engineer,65000,USD,65000,AE,100,AE,S
2022,2022,SE,FT,3D Computer Vision Researcher,10000,USD,10000,CA,50,AL,S
1485,2023,MI,FT,Machine Learning Engineer,50000,USD,50000,AM,0,AM,S
...,...,...,...,...,...,...,...,...,...,...,...
3701,2021,SE,FT,Research Scientist,50000,USD,50000,FR,100,US,S
3721,2020,SE,FL,Computer Vision Engineer,60000,USD,60000,RU,100,US,S
3735,2021,EN,PT,AI Scientist,12000,USD,12000,BR,100,US,S
3752,2020,EN,FT,Data Scientist,105000,USD,105000,US,100,US,S
